File: src-tauri/src/menu.rs
use tauri::{
	menu::{MenuBuilder, MenuItemBuilder, PredefinedMenuItem, SubmenuBuilder},
	App, Manager,
};
use tauri_plugin_opener::OpenerExt;

pub fn manage_menu(app: &App) -> Result<(), Box<dyn std::error::Error>> {
	#[allow(unused_mut)]
	let mut menu = MenuBuilder::new(app);

	#[cfg(target_os = "macos")]
	{
		let app_entry = SubmenuBuilder::new(app, &app.package_info().name.clone())
			.item(&MenuItemBuilder::new("About").build(app)?)
			.item(&PredefinedMenuItem::separator(app)?)
			.item(&MenuItemBuilder::new("Preferences...").build(app)?)
			.item(&PredefinedMenuItem::separator(app)?)
			.item(&MenuItemBuilder::new("Services").build(app)?)
			.item(&PredefinedMenuItem::separator(app)?)
			.item(&MenuItemBuilder::new("Hide").build(app)?)
			.item(&MenuItemBuilder::new("Hide Others").build(app)?)
			.item(&MenuItemBuilder::new("Show All").build(app)?)
			.item(&PredefinedMenuItem::separator(app)?)
			.item(&MenuItemBuilder::new("Quit").build(app)?)
			.build()?;

		menu = menu.item(&app_entry);
	}

	#[allow(unused_mut)]
	let mut file = SubmenuBuilder::new(app, "File")
		.item(
			&MenuItemBuilder::new("Add Channel...")
				.accelerator("CmdOrCtrl+N")
				.build(app)?,
		)
		.item(&MenuItemBuilder::new("Open").build(app)?)
		.item(&MenuItemBuilder::new("Open Channel").build(app)?)
		.item(
			&MenuItemBuilder::new("Archive")
				.accelerator("CmdOrCtrl+Backspace")
				.build(app)?,
		)
		.item(
			&MenuItemBuilder::new("Unarchive")
				.accelerator("Shift+CmdOrCtrl+Backspace")
				.build(app)?,
		)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(&PredefinedMenuItem::close_window(app, "Close".into())?);

	#[cfg(not(target_os = "macos"))]
	{
		file = file
			.item(&MenuItemBuilder::new("Options...").build(app)?)
			.item(&PredefinedMenuItem::separator(app)?);
	}

	let edit = SubmenuBuilder::new(app, "Edit")
		.item(&PredefinedMenuItem::undo(app, "Undo".into())?)
		.item(&PredefinedMenuItem::redo(app, "Redo".into())?)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(&PredefinedMenuItem::cut(app, "Cut".into())?)
		.item(&PredefinedMenuItem::copy(app, "Copy".into())?)
		.item(&PredefinedMenuItem::paste(app, "Paste".into())?)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(&PredefinedMenuItem::select_all(app, "Select All".into())?)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(
			&MenuItemBuilder::new("Find")
				.accelerator("CmdOrCtrl+F")
				.build(app)?,
		)
		.build()?;

	let view = SubmenuBuilder::new(app, "View")
		.item(
			&MenuItemBuilder::new("Show New")
				.accelerator("Alt+CmdOrCtrl+N")
				.build(app)?,
		)
		.item(
			&MenuItemBuilder::new("Show Archived")
				.accelerator("Alt+CmdOrCtrl+E")
				.build(app)?,
		)
		.item(
			&MenuItemBuilder::new("Show All")
				.accelerator("Alt+CmdOrCtrl+A")
				.build(app)?,
		)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(
			&MenuItemBuilder::new("History")
				.accelerator("CmdOrCtrl+Y")
				.build(app)?,
		)
		.item(&PredefinedMenuItem::separator(app)?)
		.item(&PredefinedMenuItem::fullscreen(app, "Fullscreen".into())?)
		.build()?;

	let window = SubmenuBuilder::new(app, "Window")
		.item(&PredefinedMenuItem::minimize(app, "Minimize".into())?)
		// .item(MenuItem::Zoom.into())
		.item(&PredefinedMenuItem::separator(app)?)
		.item(
			&MenuItemBuilder::new("Videos")
				.accelerator("Alt+CmdOrCtrl+1")
				.build(app)?,
		)
		.item(
			&MenuItemBuilder::new("Channels")
				.accelerator("Alt+CmdOrCtrl+2")
				.build(app)?,
		)
		.build()?;

	let help = SubmenuBuilder::new(app, "Help")
		.item(&MenuItemBuilder::new("Get Started").build(app)?)
		.item(&MenuItemBuilder::new("Learn More").build(app)?)
		.build()?;

	let menu_built = menu
		.item(&file.build()?)
		.item(&edit)
		.item(&view)
		.item(&window)
		.item(&help)
		.build()?;

	app.set_menu(menu_built)?;

	app.app_handle()
		.on_menu_event(|app, event| match event.id().0.as_str() {
			"Learn More" => {
				let url = "https://github.com/probablykasper/kadium";
				app.app_handle()
					.opener()
					.open_url(url, None::<&str>)
					.unwrap();
			}
			_ => {}
		});

	Ok(())
}

File: src/routes/menu.ts
import { goto } from '$app/navigation'
import { settingsOpen, viewOptions } from '$lib/data'
import { Menu, Submenu, type SubmenuOptions } from '@tauri-apps/api/menu'
import { openUrl } from '@tauri-apps/plugin-opener'
import { show_get_started } from './+layout.svelte'

export const menu_actions: Partial<
	Record<'Find' | 'Open' | 'Open Channel' | 'Archive' | 'Unarchive', () => void>
> = {}

export async function create_menu() {
	const app_menu: SubmenuOptions = {
		text: 'Kadium',
		items: [
			{
				item: {
					About: null,
				},
			},
			{ item: 'Separator' },
			{
				text: 'Preferences...',
				id: 'Preferences...',
				accelerator: 'cmdOrControl+,',
				action() {
					settingsOpen.set(true)
				},
			},
			{ item: 'Separator' },
			{ item: 'Services' },
			{ item: 'Separator' },
			{ item: 'Hide' },
			{ item: 'HideOthers' },
			{ item: 'ShowAll' },
			{ item: 'Separator' },
			{ item: 'Quit' },
		],
	}
	const file_menu: SubmenuOptions = {
		text: 'File',
		items: [
			{
				text: 'Add Channel...',
				accelerator: 'cmdOrControl+N',
				action() {
					goto('/channels?add', { replaceState: true })
				},
			},
			{
				text: 'Open',
				action: () => menu_actions.Open?.(),
			},
			{
				text: 'Open Channel',
				action: () => menu_actions['Open Channel']?.(),
			},
			{
				text: 'Archive',
				accelerator: 'CmdOrCtrl+Backspace',
				action: () => menu_actions.Archive?.(),
			},
			{
				text: 'Unarchive',
				accelerator: 'Shift+CmdOrCtrl+Backspace',
				action: () => menu_actions.Unarchive?.(),
			},
			{ item: 'Separator' },
			{
				text: 'Options...',
				id: 'Preferences...',
				accelerator: 'cmdOrControl+,',
				action() {
					settingsOpen.set(true)
				},
			},
			{ item: 'Separator' },
			{ item: 'CloseWindow' },
		],
	}
	const edit_menu: SubmenuOptions = {
		text: 'Edit',
		items: [
			{ item: 'Undo' },
			{ item: 'Redo' },
			{ item: 'Separator' },
			{ item: 'Cut' },
			{ item: 'Copy' },
			{ item: 'Paste' },
			// #[cfg(not(target_os = "macos"))]
			{ item: 'Separator' },
			{ item: 'SelectAll' },
			{ item: 'Separator' },
			{
				text: 'Find',
				accelerator: 'CmdOrCtrl+F',
				action() {
					menu_actions.Find?.()
				},
			},
		],
	}
	const view_menu: SubmenuOptions = {
		text: 'View',
		items: [
			{
				text: 'Show New',
				accelerator: 'Alt+CmdOrCtrl+N',
				action() {
					goto('/', { replaceState: true })
					viewOptions.update((v) => {
						v.show_all = false
						v.show_archived = false
						return v
					})
				},
			},
			{
				text: 'Show Archived',
				accelerator: 'Alt+CmdOrCtrl+E',
				action() {
					goto('/', { replaceState: true })
					viewOptions.update((v) => {
						v.show_all = false
						v.show_archived = true
						return v
					})
				},
			},
			{
				text: 'Show All',
				accelerator: 'Alt+CmdOrCtrl+A',
				action() {
					goto('/', { replaceState: true })
					viewOptions.update((v) => {
						v.show_all = true
						v.show_archived = false
						return v
					})
				},
			},
			{ item: 'Separator' },
			{
				text: 'History',
				accelerator: 'CmdOrCtrl+Y',
				action() {
					goto('/history', { replaceState: true })
				},
			},
			{ item: 'Separator' },
			{ item: 'Fullscreen' },
		],
	}
	const window_menu: SubmenuOptions = {
		text: 'Window',
		items: [
			//
			{ item: 'Minimize' },
			{ item: 'Maximize' },
			{ item: 'Separator' },
			{
				text: 'Videos',
				accelerator: 'Alt+CmdOrCtrl+1',
				action() {
					goto('/', { replaceState: true })
				},
			},
			{
				text: 'Channels',
				accelerator: 'Alt+CmdOrCtrl+2',
				action() {
					goto('/channels', { replaceState: true })
				},
			},
		],
	}
	const help_menu = await Submenu.new({
		text: 'Help',
		items: [
			{
				text: 'Get Started',
				action() {
					show_get_started.set(true)
				},
			},
			{
				text: 'Learn More',
				action() {
					openUrl('https://github.com/probablykasper/kadium')
				},
			},
		],
	})
	const menu = await Menu.new({
		items: [app_menu, file_menu, edit_menu, view_menu, window_menu, help_menu],
	})
	menu.setAsAppMenu()
	// https://github.com/tauri-apps/tauri/issues/12652
	help_menu.setAsHelpMenuForNSApp()
}
